src/pubtools/sign/clients/registry.py

```diff
@@ -0,0 +1,168 @@
+        return self._session
+
+    def resolve_authentication(self, image_reference: str) -> Tuple[str, str]:
+        """Resolve authentication for the given image reference.
+
+        Configuration files for container authentication are searched in the
+        first place. If there's no match found in those, the username and
+        password provided to the client are used.
+
+        Args:
+            image_reference (str): Image reference to resolve authentication for.
+        Returns:
+            Tuple[str, str]: Username and password for authentication.
+        """
+        parsed = urlparse(image_reference)
+        if not parsed.scheme:
+            parsed = urlparse(f"docker://{image_reference}")
+        registry = parsed.netloc
+        existing_auth_files = []
+        auth_files = AUTH_FILES if not self.auth_file else [self.auth_file] + AUTH_FILES
+        for af in auth_files:
+            if os.path.exists(os.path.expandvars(af)):
+                existing_auth_files.append(os.path.expandvars(af))
+        for eaf in existing_auth_files:
+            parsed_conf = json.load(open(os.path.expandvars(eaf)))
+            parsed_auths = parsed_conf.get("auths")
+            if registry in parsed_auths:
+                auth = (
+                    base64.b64decode(parsed_auths.get(registry)["auth"].encode("utf-8"))
+                    .decode("utf-8")
+                    .split(":")
+                )
+                break
+        else:
+            auth = (self.username, self.password)
+        return auth
+
+    def authenticate_to_registry(self, image_reference: str, auth_header: str) -> str:
+        """Ask for auth token based on given auth header.
+
+        Ask for authentication token based on the given auth header.
+
+        Args:
+            image_reference (str): Image reference to resolve authentication for.
+            auth_header (str): Authentication header from the registry.
+        Returns:
+            str: Authentication token.
+        """
+        _, _, value = auth_header.partition("Bearer")
+        items = parse_http_list(value)
+        opts = parse_keqv_list(items)
+        unparse_parts = [
+            "https",
+            opts["realm"].replace("https://", "").split("/", 1)[0],
+            opts["realm"].replace("https://", "").split("/", 1)[1],
+            "",
+            urlencode({"service": opts["service"], "scope": opts["scope"]}),
+            "",
+        ]
+        auth_url = urlunparse(unparse_parts)
+        username, password = self.resolve_authentication(image_reference)
+        response = self.session.get(auth_url, auth=(username, password))
+        if response.status_code == 200:
+            return response.json().get("token")
+        else:
+            response.raise_for_status()
+
+    def check_container_image_exists(
+        self, image_reference: str, auth_token: Optional[AuthTokenWrapper] = None
```

Comment:
Is the user expected to supply the auth token? Shouldn't the authentication be handled by this class? In what scenario would be this functionality desirable?

Reply:
yes, you should provide AuthTokenWrapper instance with either empty token or valid one. If token is empty, authentication fails, then request to registry is made to get valid auth token which is stored in authtokenwrapper.token for the foture. So after function finishes, authtokenwrapper.token is valid token ready for next call (for the case when you're running same method on multiple tags or repos).
Well, right now scope is usually returned per repo so I guess when you check different repo, it will fail and code needs to reauthenticate again, but I think it should be possible to ask registry for pull scope for all repos.

Reply:
Either you need to provide AuthTokenWrapper with valid token or with empty string and later class will replace it with obtained token. Having token as part of instance is possible but it wouldn't be thread safe - which is not super relevant as used requests.session is also not thread safe. But I think bearer token is something ephemeral and shouldn't be stored globally. Also when you store token in class instance, I think it's not possible to tell for what repository is it generated